pkg/network/ebpf/c/runtime/tracer.c

```diff
@@ -0,0 +1,943 @@
+        // so split those 2 statements to be able to log everything
+        if (!(t->saddr_h || t->saddr_l)) {
+            log_debug("ERR(read_conn_tuple.v6): src addr not set: src_l:%d,src_h:%d\n",
+                t->saddr_l, t->saddr_h);
+            return 0;
+        }
+
+        if (!(t->daddr_h || t->daddr_l)) {
+            log_debug("ERR(read_conn_tuple.v6): dst addr not set: dst_l:%d,dst_h:%d\n",
+                t->daddr_l, t->daddr_h);
+            return 0;
+        }
+
+        // Check if we can map IPv6 to IPv4
+        if (is_ipv4_mapped_ipv6(t->saddr_h, t->saddr_l, t->daddr_h, t->daddr_l)) {
+            t->metadata |= CONN_V4;
+            t->saddr_h = 0;
+            t->daddr_h = 0;
+            t->saddr_l = (__u32)(t->saddr_l >> 32);
+            t->daddr_l = (__u32)(t->daddr_l >> 32);
+        } else {
+            t->metadata |= CONN_V6;
+        }
+    }
+#endif
+
+    // Retrieve ports
+    t->sport = read_sport(skp);
+    bpf_probe_read(&t->dport, sizeof(t->dport), &skp->__sk_common.skc_dport);
+    t->dport = bpf_ntohs(t->dport);
+    if (t->sport == 0 || t->dport == 0) {
+        log_debug("ERR(read_conn_tuple.v4): src/dst port not set: src:%d, dst:%d\n", t->sport, t->dport);
+        return 0;
+    }
+
+    return 1;
+}
+
+static __always_inline void update_conn_stats(conn_tuple_t* t, size_t sent_bytes, size_t recv_bytes, u64 ts) {
+    conn_stats_ts_t* val;
+
+    // initialize-if-no-exist the connection stat, and load it
+    conn_stats_ts_t empty = {};
+    bpf_map_update_elem(&conn_stats, t, &empty, BPF_NOEXIST);
+    val = bpf_map_lookup_elem(&conn_stats, t);
+
+    // If already in our map, increment size in-place
+    if (val != NULL) {
+        if (sent_bytes) {
+            __sync_fetch_and_add(&val->sent_bytes, sent_bytes);
+        }
+        if (recv_bytes) {
+            __sync_fetch_and_add(&val->recv_bytes, recv_bytes);
+        }
+        val->timestamp = ts;
+    }
+}
+
+static __always_inline void update_tcp_stats(conn_tuple_t* t, tcp_stats_t stats) {
+    // query stats without the PID from the tuple
+    __u32 pid = t->pid;
+    t->pid = 0;
+
+    // initialize-if-no-exist the connetion state, and load it
+    tcp_stats_t empty = {};
+    bpf_map_update_elem(&tcp_stats, t, &empty, BPF_NOEXIST);
+
+    tcp_stats_t* val = bpf_map_lookup_elem(&tcp_stats, t);
+    t->pid = pid;
+    if (val == NULL) {
+        return;
+    }
+
+    if (stats.retransmits > 0) {
+        __sync_fetch_and_add(&val->retransmits, stats.retransmits);
+    }
+
+    if (stats.rtt > 0) {
+        // For more information on the bit shift operations see:
+        // https://elixir.bootlin.com/linux/v4.6/source/net/ipv4/tcp.c#L2686
+        val->rtt = stats.rtt >> 3;
+        val->rtt_var = stats.rtt_var >> 2;
+    }
+
+    if (stats.state_transitions > 0) {
+        val->state_transitions |= stats.state_transitions;
+    }
+}
+
+static __always_inline void increment_telemetry_count(enum telemetry_counter counter_name) {
```

Comment:
Thoughts on extracting the common inlined functions to something like `common.h`? There is a lot of duplication now between `runtime/tracer.c` and `prebuilt/tracer.c`.

Reply:
As long as the functions don't require any kind of offset, then yeah we can share them. I would prefer a filename more meaningful than `common.h` though. I've done this a little bit with the HTTP code you added. I didn't want to bloat this PR too much with changes that are tangential to runtime compilation.

Reply:
Yeah, maybe `tcp.h`, `udp.h` I dunno. I suggested `common.h` as a starting point so you don't have to think too much about this.